nifi-extension-bundles/nifi-github-bundle/nifi-github-extensions/src/main/java/org/apache/nifi/github/GitHubFlowRegistryClient.java

```diff
@@ -0,0 +1,633 @@
+
+        final String branch = flowLocation.getBranch();
+        final String filePath = getSnapshotFilePath(flowLocation);
+
+        final List<GHCommit> commits = repositoryClient.getCommits(filePath, branch);
+        final String latestVersion = commits.isEmpty() ? null : commits.getFirst().getSHA1();
+        return Optional.ofNullable(latestVersion);
+    }
+
+    @Override
+    public String generateFlowId(final String flowName) {
+        return flowName.trim()
+                .replaceAll("\\s", "-") // replace whitespace with -
+                .replaceAll("[^a-zA-Z0-9-]", "") // replace all other invalid chars with empty string
+                .replaceAll("(-)\\1+", "$1"); // replace consecutive - with single -
+    }
+
+    private FlowRegistryBucket createFlowRegistryBucket(final String name) {
+        final FlowRegistryPermissions bucketPermissions = new FlowRegistryPermissions();
+        bucketPermissions.setCanRead(true);
+        bucketPermissions.setCanWrite(true);
+        bucketPermissions.setCanDelete(true);
+
+        final FlowRegistryBucket bucket = new FlowRegistryBucket();
+        bucket.setIdentifier(name);
+        bucket.setName(name);
+        bucket.setPermissions(bucketPermissions);
+        return bucket;
+    }
+
+    private RegisteredFlowSnapshotMetadata createSnapshotMetadata(final GHCommit ghCommit, final FlowLocation flowLocation) throws IOException {
+        final GHCommit.ShortInfo shortInfo = ghCommit.getCommitShortInfo();
+
+        final RegisteredFlowSnapshotMetadata snapshotMetadata = new RegisteredFlowSnapshotMetadata();
+        snapshotMetadata.setBranch(flowLocation.getBranch());
+        snapshotMetadata.setBucketIdentifier(flowLocation.getBucketId());
+        snapshotMetadata.setFlowIdentifier(flowLocation.getFlowId());
+        snapshotMetadata.setVersion(ghCommit.getSHA1());
+        snapshotMetadata.setAuthor(ghCommit.getAuthor().getLogin());
+        snapshotMetadata.setComments(shortInfo.getMessage());
+        snapshotMetadata.setTimestamp(shortInfo.getCommitDate().getTime());
+        return snapshotMetadata;
+    }
+
+    private String getSnapshotFilePath(final FlowLocation flowLocation) {
+        return SNAPSHOT_FILE_PATH_FORMAT.formatted(flowLocation.getBucketId(), flowLocation.getFlowId());
+    }
+
+    private RegisteredFlowSnapshot getSnapshot(final String filePath, final String branch) throws IOException, FlowRegistryException {
+        try (final InputStream contentInputStream = repositoryClient.getContentFromBranch(filePath, branch)) {
+            return OBJECT_MAPPER.readValue(contentInputStream, RegisteredFlowSnapshot.class);
+        }
+    }
+
+    private RegisteredFlowSnapshot getSnapshot(final InputStream inputStream) throws IOException {
+        try {
+            return OBJECT_MAPPER.readValue(inputStream, RegisteredFlowSnapshot.class);
+        } finally {
+            IOUtils.closeQuietly(inputStream);
```

Comment:
I think some logging would be useful for debugging purposes

Reply:
Not sure there is anything to log here, if an exception happened it would throw IOException out of this method and propagate. Failing to close the stream isn't very meaningful.